libjcat/jcat-bt-util.c

```diff
@@ -0,0 +1,1260 @@
+	}
+
+	ed25519_sha512_sign(parsed_public_key->data,
+			    parsed_private_key->data,
+			    text->len,
+			    (const guint8 *)text->str,
+			    raw_signature->data);
+
+	/* Add the key hash prior to the signature */
+	g_byte_array_prepend(raw_signature, (const guint8 *)parsed_key_hash, 4);
+
+	encoded_signature = g_base64_encode(raw_signature->data, raw_signature->len);
+
+	/* Add text and an extra new line (need two consecutive newlines) */
+	buf = g_byte_array_new();
+	g_byte_array_append(buf, (const guint8 *)text->str, text->len);
+	g_byte_array_append(buf, (const guint8 *)"\n", 1);
+
+	/* Format the signature */
+	g_string_printf(formatted_signature,
+			"\xe2\x80\x94 %s %s\n",
+			parsed_key_name,
+			encoded_signature);
+	g_byte_array_append(buf,
+			    (const guint8 *)formatted_signature->str,
+			    formatted_signature->len);
+	return g_byte_array_free_to_bytes(buf);
+#endif
+}
+
+static const gchar *
+memrmem(const gchar *haystack, gsize haystack_size, const gchar *needle, gsize needle_size)
+{
+	/* Like the standard library function memmem, but finds the last occurrence. */
+	const gchar *p = memmem(haystack, haystack_size, needle, needle_size);
+	const gchar *q;
+	if (p == NULL)
+		return NULL;
+	for (;;) {
+		p += needle_size;
+		q = memmem(p, haystack + haystack_size - p, needle, needle_size);
+		if (q == NULL)
+			return p - needle_size;
+		p = q;
+	}
+}
+
+static gboolean
+parse_signed_note(GBytes *signed_note,
+		  GByteArray *parsed_public_key,
+		  GBytes **original_text,
+		  gchar *expected_key_name,
+		  gchar *expected_key_hash,
+		  GError **error)
+{
+#ifndef ENABLE_ED25519
+	g_set_error_literal(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    __FUNCTION__ " requires Ed25519 support but it is not enabled");
+	return FALSE;
+#else
+
+	gsize size = 0;
+	const gchar *data = g_bytes_get_data(signed_note, &size);
+	const gchar *signature_start = NULL;
+	gsize signature_size = 0;
+	const gchar *p, *end;
+	g_autofree guchar *sig_str = NULL;
+	g_autofree gchar *temp_sig = NULL;
+	g_autofree gchar *key_name = NULL;
+	gsize sig_size = 0;
+	guchar *actual_key_hash = NULL;
+
+	g_return_val_if_fail(original_text != NULL, FALSE);
+
+	/* Must be valid UTF-8 */
+	if (!g_utf8_validate(data, size, NULL)) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "signed note is not valid UTF-8");
+		return FALSE;
+	}
+
+	/* Must not contain any control characters other than newline */
+	for (p = data, end = data + size; p != end; p = g_utf8_next_char(p)) {
+		gunichar u = g_utf8_get_char(p);
+		if (u < 0x20 && u != '\n') {
+			g_set_error_literal(error,
+					    G_IO_ERROR,
+					    G_IO_ERROR_FAILED,
+					    "signed note contains invalid character");
+			return FALSE;
+		}
+	}
+
+	/* Get the signature block, which is delineated by two consecutive newline characters.
+	 * Unfortunately we need to find the last occurrence. */
+	signature_start = memrmem(data, size, "\n\n", 2);
```

Comment:
maybe a dumb question, but wouldn't g_strsplit(data, "\n\n", -1) make this code a lot simpler